### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by zlm3110561045** [gographics/imagick#136]
Hi, sorry to bother you. I was doing some testing since one of my applications, the memory is never cleared after the wand is destroyed. 

Code used to test:

func main(){

       // data image content ...
        mw, err := xcimagick.NewImagick(data)
	if err != nil {
		logger.Error(err)
		return
	}
	defer mw.Destory()

       // ...
}

func Init() (err error) {
	imagick.Initialize()
	return
}

func Destory() (err error) {
	imagick.Terminate()
	return
}

type MagicWand struct {
	mw *imagick.MagickWand
}

func NewImagick(data []byte) (m *MagicWand, err error) {
	mw := imagick.NewMagickWand()
	err = mw.ReadImageBlob(data)
	if err != nil {
		logger.Error(err)
		return
	}

	m = &MagicWand{mw: mw}
	return
}

func (m *MagicWand) GetImageFormat() string {
	return strings.ToLower(m.mw.GetImageFormat())
}

func (m *MagicWand) GetImagesBlob() []byte {
	return m.mw.GetImagesBlob()
}

func (m *MagicWand) Destory() {
	m.mw.Destroy()
}

func (m *MagicWand) ConvertGifToJpeg() (ret *MagicWand, err error) {
	ret = NewMagickWand()
	bcolor := NewPixelWand()

	for i := 1; i < int(m.mw.GetNumberImages()); i++ {
		m.mw.SetIteratorIndex(i)
		w, h, x, y, err := m.mw.GetImagePage()
		if err != nil {
			logger.Error(err)
			continue
		}

		tmp := m.mw.GetImage()
		ret.mw.NewImage(w, h, bcolor)
		ret.mw.SetImageColorspace(tmp.GetImageColorspace())
		ret.mw.CompositeImage(tmp, tmp.GetImageCompose(), true, x, y)
		tmp.Destroy()
		break
	}

	m.mw.Destroy()
	if err != nil {
		logger.Error(err)
		return
	}
	return
}

func (m *MagicWand) CompressImageJpeg(quality uint) (err error) {
	err = m.mw.SetImageCompression(imagick.COMPRESSION_JPEG)
	if err != nil {
		logger.Error(err)
		return
	}

	if quality > 0 {
		orgQ := m.mw.GetImageCompressionQuality()
		if orgQ > quality {
			err = m.mw.SetImageCompressionQuality(quality)
			if err != nil {
				logger.Error(err)
				return
			}
		}
	}

	m.mw.SetImageFormat("jpeg")
	return
}

func (m *MagicWand) GetImageSize() (w, h uint, err error) {
	if m.GetImageFormat() == "gif" {
		w, h, _, _, err = m.mw.GetImagePage()
	} else {
		w = m.mw.GetImageWidth()
		h = m.mw.GetImageHeight()
	}
	return
}

func (m *MagicWand) ResizeImage(w, h uint) (err error) {
	return m.mw.ResizeImage(w, h, imagick.FILTER_CATROM)
}

func (m *MagicWand) CropImage(w, h uint, x, y int) (err error) {
	return m.mw.CropImage(w, h, x, y)
}

func (m *MagicWand) GetImageMime() string {
	return fmt.Sprintf("image/%s", strings.ToLower(m.mw.GetImageFormat()))
}

func (m *MagicWand) GetImageOrientation() int {
	return int(m.mw.GetImageOrientation())
}

**Comment by justinfx** [comment]
Simply not having the memory cleared doesn't necessarily mean it's a memory leak. I'm not saying it isn't a leak, but Go runtime only advises the operating system that it no longer needs pages of memory. The OS doesn't have to reclaim it until it needs it. 

The question is whether your memory continues to grow and grow when you rerun the same operations. Is that the case? Or does it always just stay at the same max RSS? 

If it is indeed growing constantly, can you provide the smallest valid runnable program to reproduce it?

**Comment by zlm3110561045** [comment]
Hi, sorry to bother you again. I tried to find the problem code, and the memory leak seemed to appear in the following code through 100 GIF tests

mw := imagick.NewMagickWand()
err = mw.ReadImageBlob(data)
if err != nil {
	logger.Error(err)
	return
}
mw.Destory()

**Comment by justinfx** [comment]
Alright. So you are saying reading 100 gifs in a row, and creating and destroying a wand each time leads to a continuously growing resident memory size on each image? What RSS does the memory start at and what does it climb to after 100?

Also what version of ImageMagick and what platform are you running?

**Comment by advancedlogic** [comment]
as a suggestion, in any case,  I would put defer mw.Destroy() after mw creation

**Comment by tynia** [comment]
` mw := imagick.NewMagickWand()
  defer mw.Destroy()
  ctx, err = cache.Get(ImageID)
  if err != nil {
      logging.Error("try to get original image failed, error = %s", err.Error())
      return
  }

  errRead := mw.ReadImageBlob(ctx)
  if errRead != nil {
      logging.Error("failed to read image from byte, error = %s", errRead.Error())
      err = e.NewWError(e.ERR_CODE_SYS, "failed to read image from byte")
      return
  }

  mw.ResetIterator()
  orientation := mw.GetImageOrientation()
  if orientation > 1 {
      err := mw.AutoOrientImage()
      ...
  }
  //call another function, main code in the function as below
｛
     mw.ResizeImage(w, h, imagick.FILTER_LANCZOS)
     mw.SetImageCompressionQuality(95)
     mw.SetImageFormat("jpg")
     mw.StripImage()
  ｝
   ...
`

I run it about 1011 times，which process a image(570kb) from  2000x2000 to 400x400, and the memory the program hold more than 1gb.

and if i comment those code, just return the original image content, the memory will stay less than 100mb (my program hold 80mb when starting)

ImageMagick version: ImageMagick-7.0.7-7
the imagick version is the 3.1.1

**Comment by justinfx** [comment]
Can you please reformat this code into an exact runnable reproduction of your problem? Because what you have here is ambiguous. Are you doing this in a for loop? Are you using defer in a for loop? I would need to see an exact 100% example of what you are doing.

**Comment by tynia** [comment]
ok，you just try to read a image blob, and you will find the problem,
```
func Resize(id string, w, h int) ([]byte, error) {
   mw := imagick.NewMagickWand()
   defer mw.Destroy()
   ctx, err := cache.Get(id) // you can read data from a local pic and store in []byte
   mw.ReadImageBlob(ctx)
   return
}
```

i just do this in my test request handler, the ctx is instance of []byte, which hold the original image data got from object storage service. you can read a local pic into bytes and call **ReadImageBlob**

 i'm sorry about this i cannot apply a exact number the memory leak ,the number between 250-300 byte leak every run. and i find the RES will climb to more than 10 times of the original image data when i called the "ReadImageBlob". 

i send 200 requests in the same time  and the item of RES in top can climb to 5.570g  while the original length of ctx is 1217542 byte.

**Comment by justinfx** [comment]
Here is a complete test program. I used the same ImageMagick version as you had listed, using go 1.9 on OSX. I did not see a memory leak. This also does not contain any of your custom application logic such as caches. 

```go
package main

import (
	"fmt"
	"os"
	"time"

	"gopkg.in/gographics/imagick.v3/imagick"
)

func main() {
	fmt.Printf("PID: %d\n", os.Getpid())
	fmt.Println("Creating image")
	b := create()

	fmt.Println("Looping...")
	const N = 1200
	for i := 0; i < N; i++ {
		fmt.Printf("read %d/%d\n", i+1, N)
		read(b)
		// read(create()) // create new image each loop
	}

	fmt.Println("Done looping. Sleeping.")
	time.Sleep(1 * time.Minute)
}

func create() []byte {
	in := imagick.NewMagickWand()
	defer in.Destroy()
	in.SetSize(2000, 2000)
	in.ReadImage("magick:logo")
	in.SetImageFormat("jpg")
	in.SetImageCompressionQuality(95)
	return in.GetImageBlob()
}

func read(b []byte) {
	in := imagick.NewMagickWand()
	in.ReadImageBlob(b)
	in.Destroy()
}
```

Are you able to see a leak with this program? Are you able to modify it to cause the leak?

**Comment by tynia** [comment]
yeah, your program seems absolutely okay.

however, can you read it parallel?
```
func main() {
	fmt.Printf("PID: %d\n", os.Getpid())
	fmt.Println("Creating image")
	b := create()

	fmt.Println("Looping...")
	const N = 1200
	for i := 0; i < N; i++ {
		fmt.Printf("read %d/%d\n", i+1, N)
		**go** read(b)
		// read(create()) // create new image each loop
	}

	fmt.Println("Done looping. Sleeping.")
	time.Sleep(1 * time.Minute)
}
```

**Comment by tynia** [comment]
i believe you have seen the RES the program holds didn't decrease even thought days later

**Comment by justinfx** [comment]
Sorry for the delay. I have only just now gotten time to look at this again. I've tested the program with the `read()` running in a new goroutine for each loop (on my OSX laptop). I do see some memory remaining allocated beyond what the Go runtime reports as being part of its heap allocation. It is in the 120-150mb range and does not decrease for the life of the program. Setting "MAGICK_DEBUG=Trace" shows a good deal of Destroy operations happening, but maybe it is leaking somewhere in the ImageMagick C layer. 

When I was using this binding in a long running resize and cache service, I had always been running it with a bounded number of image processing workers. Being a cgo binding, the call into C requires an OS thread, so I always wanted to have a handle on the contention for calling into ImageMagick. This is probably why I never noticed any leak issues. It may be a problem with the concurrent threads in C ImageMagick.

```go
	const NReaders = 4
	const NLoops = 1200

	sem := make(chan struct{}, NReaders)

	fmt.Println("Looping...")
	for i := 0; i < NLoops; i++ {
		i := i
		go func() {
			sem <- struct{}{}
			fmt.Printf("read %d/%d\n", i+1, NLoops)
			read(b)
			<-sem
		}()
	}
``` 

In this modified program the semaphore is only running 4 calls at any one time. Running this program shows me 13 threads are being used by the entire program, and a memory usage of ~21MB. If I run the original unbounded program I get ~110 threads and the ~150MB usage. 

Does your own program use unbounded calls into ImageMagick? I feel it may be reasonable to expect high resource consumption when using a cgo library with number of concurrent calls being dictated by the client connections.

**Comment by tynia** [comment]
requests to my service usually gather in the same period of time.
in order to digit out the resources it needs, i have to run a press test to it. so i close the channel to test how many requests it can process at the same time, normally the capacity of channel is set to 200.
during testing, i find it goes well when request comes one by one, but leaks when processing requests parallel. i would like to believe that it is the problem in GC wait. however, it make no difference when i try to force to GC, as well as sync.Pool.

i try to read some code in MagickWand and MagickCore, it seems use a **reference** to reduce the number of time to copy. i guess it may raise the problem of leak when bytes in GO passed to C API and it remembered a wrong number with it. (my guess may be wrong because it runs well one by one).

thanks for you work.

happay new year to you

**Comment by jiecongwang** [comment]
We probably are running into same issues, is there any update?  Is this issue https://github.com/ImageMagick/ImageMagick/issues/904   relevant?

**Comment by huhr** [comment]
I have the same problem,  my program is a HTTP Server for
 cropping image  as serval given sizes，I use MagickWand.Clone to create new MagickWand instance  and  crop the  for each size parallelly.  I find as my server running on，the processing time is slowly increasing until too many requests blocked

**Comment by justinfx** [comment]
@huhr that sounds like a different problem. What do you mean by too many request being blocked? The original problem talks about memory not being released, which may or may not be normal if the host isn't under too much memory pressure. When you say requests are blocked, what does that look like?

**Comment by huhr** [comment]
my program is a HTTP Server, I first start server, the average processing time for each request is about 700ms, as runing on, the average processing time climb to seconds or even more. since program has no request rate limit or timeout yet，at some point, the processing time has a sudden increase and all request coming without process done. 
Strangely，the requests per second is  stable all the time and usage of cpu  or memory  is not vely high，I had run it on a 48 hyperthread physical maching with MAGICK_THREAD_LIMIT set to be about 12.

**Comment by justinfx** [comment]
Yea, this does sound like a different problem than this originally reported issue. This issue is suggesting there is a memory leak and that the memory continues to climb for the run of the program. I have no idea why your http server appears to serve requests more slowly over time. I recommend you do some profiling and find out where the request is spending the time. Or have you absolutely narrowed it down to a call into ImageMagick? If you are sure it's ImageMagick, maybe open another issue with the results of your profiling and a reproduction of the problem.

**Comment by aclowkey** [comment]
@justinfx Hi, I'm new to the Go world, and especially CGo.  This library seemed like a good interface to use the MagickWand functionality. Now however, I'm not so sure. 

Can you simply describe the implications of this? If I deploy this app to production will it crash after regular use? i.e. same amount of requests without any unusual peaks.

**Comment by justinfx** [comment]
@aclowkey, that is a very general question. When I was using this library in my image caching server, I experienced no issues. But it seems, according to some of the issue reports, that there are certain environments and use cases that are giving users problem. I've not pinpointed any memory leaks from this report. Though I am not claiming imagick is 100% bug free and you will never have a problem.

Full disclosure, I am not actually using this library actively anymore. I switched my image cache server over to using my bindings for OpenImageIO: https://github.com/justinfx/openimageigo
So I am not really putting this project through its paces anymore and I try to just offer maintenance to its active users.

**Comment by justinfx** [comment]
@aclowkey one more afterthought. I mentioned previously that my own use case was to have bounded works for performing the imagemagick conversions, instead of having every request concurrently do image creations. My tests did show an increase in memory usage if you don't control the amount of concurrent conversions. To me, it makes no sense to allow 200 imagemagick conversions to launch at the same time through concurrent http requests. That would explain the high memory consumption. Also the original post didn't suggest there was actually a crash. Only a high usage of reported ram. I would only be concerned if the machine was actually going into swap.

**Comment by Eriken** [comment]
Ok.
I have been having similar issues to what has been described here. And after some procrastination I decided to do some digging. 

I use these bindings in three services at our shop. And my test was based on a service that handles resize requests for a web frontend.

This service created MagickWands and PixelWands for each request and destroyed them after use. But after reading some of @justinfx comments I realised that making a pool of these for reuse and resource management might be a good idea. My solution was to populate workers embedding both wands on a channel and pull and release them as needed.

I also realised there was a function called [Clear()](https://github.com/gographics/imagick/blob/master/imagick/types/clearer.go) for wands. So instead of creating new ones and destroying them each time I decided to reuse them by clearing them after each use.

After running long iterations of tests with this implementation I was still seeing continuous memory consumption. At the time I was using IM version `6.9.9-51`. After decrementing the patch version down to version `6.9.5-10` I was finally seeing more stable memory consumption.

So I am now running a pool of reusable worker wands with IM version `6.9.5-10`, that seem stable and with lower memory footprint, roughly 4  GB(had up to 24 GB prior to this) with roughly 100 reqs per minute per node, we run 5 nodes.

I am honestly not sure what causes this, I have not investigated the diffs. But if you are struggling with increasing memory consumption this might be worth a try.

@justinfx 
Thanks for your effort! 
If more users can confirm that reuse and limitation of wands help with memory consumption then it might be worth to update the README with a more opinionated implementation example. I would be happy to contribute. 

Thanks again.

**Comment by justinfx** [comment]
Thanks for this information and the time you put into testing. Personally I don't end up having time to test these things, and the motivation can be quite low because most the time the problem is related to ImageMagick and not the bindings.

I don't recall suggesting keeping the wands around and reusing them. But I have suggested using bounded image workers so that you don't end up with as many ImageMagick conversions as you have http requests. Does it make a difference in memory consumption when you use the worker pool but create and destroy wands on every conversion? Or is it just a matter of the ImageMagick version and the worker pool?

**Comment by Eriken** [comment]
> Does it make a difference in memory consumption when you use the worker pool but create and destroy wands on every conversion?
I did not test this.

> Or is it just a matter of the ImageMagick version and the worker pool?
Decrementing the IM version and capping the workers seemed to have the most effect.

Nodes have been running stable for a day now.

**Comment by Eriken** [comment]
Ok, time for another update.

So after locking our services to this one release (6.9.5-10) I decided it was time to move to IM 7 (7.0.8-12) as we were seeing too many issues not applying patches.

As I predicted this resulted in increased and growing memory consumption. We run these services within cgroups so they are just killed when they surpass as certain memory threshold, but the behaviour was obiously not optimal.

Our services processes all kinds of image formats so I tried to narrow the format scope in order to find wich format was causing this growing memory consumption. The first test was JPEG, and suprisingly enough the behaviour accord with JPEG.

We run the service in a container running Ubuntu Bionic (18.04) and used the libjpeg-dev package as a decoder. I realised this package used the 8c version og libjpeg, which dates back to 2011, the newest version is 9c.

Anyway, I decided to try out libjpeg-turbo8-dev version 1.5.2, which is from 2017. And even if the memory print is higher than we had with IM 6 it does not seem to grow anymore. High but stable.

Another point to add is that we added the following thresholds to the `policy.xml`:
```
<policy domain="resource" name="memory" value="1GiB"/>
<policy domain="resource" name="map" value="2GiB"/>
```

This should only prevent very spicky behaviour as images that exceeds these limits will be written to disk, not memory.

Hope this can be helpful. ImageMagick has a lot of settings and dependencies, mastering those seem just as important as using the API correctly.

**Comment by justinfx** [comment]
Thanks for that info. Hopefully it will help others that are doing long running image servers.
I've not been running imagick in production for a while now so I haven't had to look into this problem. My current image server uses my [OpenImageIO] (https://github.com/justinfx/openimageigo) bindings. It doesn't have as many "tweaks" to master for memory consumption. Just some usage patterns around whether you want it to use the ImageCache or not.

**Comment by Eriken** [comment]
Thanks for linking and maintaining said bindings @justinfx . Might be a library to hack with some day.

**Comment by junedev** [comment]
Unfortunately I cannot confirm the fix that @Eriken wrote about. We are also running ImageMagick 7.0.8-12 on Ubuntu 18.04 (in a Docker container) with libjpeg-turbo8-dev and we also activated the memory limit. Nevertheless we still have a memory leak that leads to the service being killed by the OS just as @Eriken explained.

**Comment by Eriken** [comment]
@justinfx : sadly I think @junedev is right.
After propagating my changes to all our imagemagick based services it seems like there are still some leaks, and services are killed by the kernel due to memory abuse.

I am not knowledgable with cgo, but I would be happy to contribute with testing and other insights.

**Comment by justinfx** [comment]
Thanks @Eriken
I don't have a lot of spare time to dig into this right now, and like I had mentioned I am not using the binding in production, so any help from others on the project is very welcome.

**Comment by Eriken** [comment]
@justinfx 
~Last 24 hours: memory consumtion is high, but stable. I have 6 workers both embedding a wand and a pixelwand. See graph below:
![memory graph](https://i.imgur.com/bsOOCvs.png)

**Comment by justinfx** [comment]
Which memory stat is that a sample of? How it is collected? It does look
similar to my image server in the sense that is holds onto a chunk and
floats up and down around that. Now I wonder if you were to run some other
process on that host that just tries to eat 8gb or something like that,
would your ImageMagick process release memory back to the os? Or would one
of the processes get OOM killed?

**Comment by Eriken** [comment]
The first graph is rendered in grafana using cadvisor stats.

This is the heap profile:
![heap](https://i.imgur.com/70QXKQZ.png)
So almost all of the memory allocated in first graph must be cgo as pprof does not show cgo allocations.

They do not get killed by the kernel at the moment because 14 GB is tolerable.
Previously they have been killed by the kernel when they reached around 20 GB though. But that was with the fault `Clear()` implementation.

I could try to lower the amount of workers and see if memory allocated dropped accordingly?

**Comment by justinfx** [comment]
In both imagick and OpenImageiGO, I noticed a correlation between the
number of workers and the minimum memory level. There was just an inherent
amount of memory that it held onto. But it wasn't constantly growing. There
was a point where it settled and would just pivot up and down around that
number. And similarly, the memory was all held in C. I was wondering about
the behaviour of periodically locking and calling
Terminate()/Initialize()  to see if ImageMagick will clean itself up.

**Comment by justinfx** [comment]
@tczf1128 it is a leak only if those workers are consistently growing in memory over time. Just holding  1-2gb doesn't necessarily mean a leak.

**Comment by tczf1128** [comment]
Can I limit the memory used? and why so much memory held by imagick

**Comment by justinfx** [comment]
You would have to ask the ImageMagick maintainers about that. It just comes down to the internal cache and how it frees resources over time. You can look at setting resource limits either through the API or your environment variables:

https://www.imagemagick.org/script/resources.php#environment
https://godoc.org/gopkg.in/gographics/imagick.v3/imagick#MagickWand.SetResourceLimit

Also be aware that simply seeing a high constant memory value doesn't mean you are leaking memory. Go holds on to memory from the OS until it really asks for it to be returned. So if you aren't swapping or encountering OOM errors, then you should be fine.

**Comment by tczf1128** [comment]
Thanks

**Comment by tczf1128** [comment]
I find it not works when I set the policy.xml in /usr/local/etc/ImageMagick-7

`<policy domain="resource" name="memory" value="1GiB"/> 
  <policy domain="resource" name="map" value="0B"/> 
  <policy domain="resource" name="disk" value="0B"/> 
  <policy domain="resource" name="file" value="0"/> 
`

**Comment by joydchh** [comment]
Is there new updates on this? We also met this problem in our production release.

**Comment by justinfx** [comment]
@joydchh no one is looking into this as far as I am aware. Are you seeing a steady increase in memory usage over time until it runs out of memory? Or are you seeing just a large chunk of memory being held?

**Comment by joydchh** [comment]
> @joydchh no one is looking into this as far as I am aware. Are you seeing a steady increase in memory usage over time until it runs out of memory? Or are you seeing just a large chunk of memory being held?

A steady increase in memory usage over time util it runs out of memory. 
Also it could be reproduced by the way tynia provided above, and the memory just could not be released.

**Comment by jeevi-cao** [comment]
> > @joydchh no one is looking into this as far as I am aware. Are you seeing a steady increase in memory usage over time until it runs out of memory? Or are you seeing just a large chunk of memory being held?
> 
> A steady increase in memory usage over time util it runs out of memory. Also it could be reproduced by the way tynia provided above, and the memory just could not be released.

i find this question too

**Comment by justinfx** [comment]
@jeevi-cao  there is a lot of information on this thread. You would need to be more specific about what you have tried. To me, it seems that a big contributor to memory growth is when you have unbounded goroutines doing a lot of concurrent ImageMagick conversions. My thinking is that because it is cgo, and each call into cgo needs an real OS thread, doing very large amounts of concurrent calls will result in many OS threads needing to be started. And then it is likely that ImageMagick is doing some kind of thread-local storage caching, which would lead to large memory growth. This is why, when I used to use this library, I usually didn't see huge memory growth. I was using bounded concurrency, where there could only be a max number of concurrent image conversions at any time (such as 8). You can see from the reported memory profiles in this thread that most the memory is being held in C and not Go. A good indicator that it relates to ImageMagick caches is if Terminate is called, and we see the memory freed. That is why I thought it also might be a workaround to use a `RWLock` where all the conversion goroutines use the "read lock", and occasionally the "write lock" is acquired to terminate and initialize ImageMagick again.

I still think if anyone really wants to get to the bottom of this, a discussion with the ImageMagick devs would be a good idea, to better understand the implication of using 200 OS threads in a long running application, and what happens within ImageMagick with thread-local storage. I think its a bad idea to let your program do that in the first place, and you should be using a bounded number.

One other thing to try, for anyone investigating this,   is to see if locking the goroutine to the OS thread makes any difference to the memory growth:

https://pkg.go.dev/runtime#LockOSThread

```go
func MyImageHandlers() {
    runtime.LockOSThread()
    defer runtime.UnlockOSThread ()
    doImageMagickThing()
    doAnotherImageMagickThing()
}
```

It could be possible that without this call, in an application with tons of concurrent ImageMagick calls , different OS threads might be used in the same goroutine, and maybe that causes some kind of leak if ImageMagick is using thread-local storage and you are going back and forth with wand operations?

**Comment by sinni800** [comment]
> @jeevi-cao
> One other thing to try, for anyone investigating this, is to see if locking the goroutine to the OS thread makes any difference to the memory growth:
> 
> https://pkg.go.dev/runtime#LockOSThread
> 
> ```go
> func MyImageHandlers() {
>     runtime.LockOSThread()
>     defer runtime.UnlockOSThread ()
>     doImageMagickThing()
>     doAnotherImageMagickThing()
> }
> ```
> 
> It could be possible that without this call, in an application with tons of concurrent ImageMagick calls , different OS threads might be used in the same goroutine, and maybe that causes some kind of leak if ImageMagick is using thread-local storage and you are going back and forth with wand operations?

I'm currently doing a thing where i need to use this to make 4.6 million thumbnails, and despite me Destroy()ing EVERY imagick object, I just keep amassing more RAM usage. the course of a few hundred thousand images it goes to like 60+ GB of usage.

Essentially, my workflow is this: 

```
load image into imagick object 1
create another, object 2
resize object 1
composite object 1 onto 2 
```

I've taken good care that destroy is called, and in an attempt to maybe reduce the leak, I've started to reusing "object 1", and was right now working on reusing "object 2" too (by throwing a Clear() onto them on each use)

I've used your tip and I am now locking goroutines to OS threads... Fingers crossed.

**Comment by justinfx** [comment]
@sinni800 in addition to locking to the OS thread, are you using bounded or unbounded goroutines? As mentioned before, I have found that because cgo will launch a new OS thread every time a cgo call is blocked, you can end up with many many threads in a highly concurrent usage of imagick. I've found memory usage much more predictable when I use a semaphore or a worker pattern to limit the max concurrent ImageMagick calls.

**Comment by sinni800** [comment]
> 

I currently use a fixed amount of goroutines that work off of a channel that is being filled with mysql results (I basically make thumbnails off of images, fed from a database of gallery rows)

Currently, that amount is 3, so not even *that* much. I am leaking less memory by locking to OS Threads, but it's still not perfect, I am using more and more RAM over time. Pprof shows nothing so that ram isn't locked in Go memory, it's probably the ImageMagick memory. 

I also went back to creating a new MagickWand for each image, because somehow I was getting slowdowns the longer the app ran when I reused them after Clear()ing them. 

Basically I have my library here: https://git.metalgearsonic.de/sinni800/booruwerks/src/branch/master/booruimage/magick2/imageimagick.go

I first run GetImageF, then do one RescaleRatio, then EncodeFileName, then Close() on both the first image gotten from GetImageF, then the one gotten from RescaleRatio (that writes a second instance composing the first image onto it) 

Do you see, on first look, anything that I am missing? 

```
img, _, err = magick.GetImagef(filepath.Join(*imagedir, p.Directory, p.Image))
timg = img.RescaleRatio(uint(*thumb1size))
err = timg.EncodeFileName(filepath.Join(*thumbdir, p.Directory, "thumbnail_"+fname+".jpg"), uint(*jpegquality))
timg.Close()
timg = nil
img.Close()
img = nil
```

is the basic runaround here. Hmm.

**Comment by justinfx** [comment]
@sinni800 yea I don't know. There is a bit of code there in that source and so it would have to be narrowed down to the smallest repro possible. And then it would have to be looked at with a C memory profiler to figure out where all that memory might be within ImageMagick.
I'm not counting out the possibility of some kind of oversight on the imagick binding side. Also testing against the latest supported ImageMagick7 w/ the v3 bindings would be a good idea as well.

**Comment by sinni800** [comment]
@justinfx I don't think it's that much code, but I threw the functions out into a separate runnable main.go anyway

https://gist.github.com/sinni800/da59b65ae4ba4acdf86f7141ec90917b

This really only has an appreciable memory leak after you do like tens of thousands images exactly like that, though.

**Comment by justinfx** [comment]
Cool. Well see if you can identify anything with a C memory profiler and a lot of image conversions?
I never use the library anymore so I don't put too much time into playing with testing it.

**Comment by sinni800** [comment]
I'm trying to get there... But for now I don't really have access to all the resources to do a long enough run so I can profile it. Though I honestly don't know what to look for. Do you think I'm doing anything wrong in freeing the resources or anything? Should Destroy clearly free everything that's used?

**Comment by justinfx** [comment]
The bindings are designed to provide the wrappers around directly cleaning up resources via the Destroy methods. And they have a fallback otherwise on relying on the garbage collector to call their finalizer functions. 
ImageMagick C API usually has "Release" functions to pair with constructor functions. So we have tried to make sure to capture those into the Go wrapper types. It's still possible we missed something. Or it's possible it's a real ImageMagick C leak. Essentially you would need to try and isolate the exact operation leading to leaks. Then we see if the C calls are being called correctly in the wrapper.
In my other binding, for OpenImageIO, I had found and fixed bugs where C strings we being allocated and not cleaned up. Something could always be missed here as well.
